src/hidra/receiver/datareceiver.py

```diff
@@ -259,6 +260,97 @@ def stop(self):
+    # slow or blocking plugins
+    try:
+        plugin_m = import_module("plugins." + plugin_name)
+        plugin = plugin_m.Plugin(plugin_config)
+        plugin.setup()
+        log.info("Loading '%s' plugin", plugin_name)
+    except Exception:
+        log.error("Could not load '%s' plugin", plugin_name, exc_info=True)
+        return
+
+    while not event.is_set():
+        try:
+            data = data_queue.get(timeout=0.5)
+        except queue.Empty:
+            continue
+
+        try:
+            [metadata, data] = data
+            plugin.process(
+                local_path=generate_filepath(target_dir, metadata),
+                metadata=metadata,
+                data=data
+            )
+        except Exception:
+            log.error("Processing data with '%s' plugin failed.",
+                      plugin_name, exc_info=True)
+
+        data_queue.task_done()
+
+    try:
+        plugin.stop()
+    except Exception:
+        log.error(
+            "Error while stopping '%s' plugin", plugin_name, exc_info=True)
+
+
+class PluginHandler:
+    def __init__(self, plugin_name, plugin_config, target_dir, log):
+        self.plugin_name = plugin_name
+        self.plugin_config = plugin_config
+        self.log = log
+
+        self.plugin_queue = queue.Queue()
+        self.plugin_event = threading.Event()
+        self.plugin_thread = threading.Thread(
+            target=run_plugin_thread,
+            args=(self.plugin_name, self.plugin_config, target_dir,
+                  self.plugin_queue, self.log, self.plugin_event),
+            daemon=True)  # stop process even if thread is still alive
+
+    def get_data_type(self):
+        return self.plugin_config.get("plugin_type", "metadata")
+
+    def start(self):
+        self.log.info("Starting '%s' plugin thread", self.plugin_name)
+        self.plugin_thread.start()
+
+    def put(self, message):
+        if self.plugin_queue.qsize() < 10000:
+            self.plugin_queue.put(message)
+        else:
+            self.log.warning(
+                "Skipping '%s' plugin because queue is full", self.plugin_name)
+
+    def stop(self, timeout=60):
+        self.log.info(
+            "Waiting for '%s' plugin thread to finish", self.plugin_name)
+
+        # wait manually because queue.join doesn't have a timeout parameter
+        while self.plugin_queue.qsize() > 0 and timeout > 0:
+            time.sleep(0.1)
+            timeout -= 0.1
+
+        if timeout <= 0:
+            self.log.error(
+                "'%s' plugin queue not empty after timeout", self.plugin_name)
+            timeout = 1  # always give the thread a second to join
+
+        self.plugin_event.set()
+
+        if self.plugin_thread:
+            self.plugin_thread.join(timeout=timeout)
+
+        if self.plugin_thread.is_alive():
+            self.log.error(
+                "'%s' plugin did not join within timeout", self.plugin_name)
+
```

Comment:
the flow logic is not entirely clear. why does the stop() function have no return value?
right now it can be called, and it seems regardless which effect it had

Reply:
`stop` is called at shutdown to give the plugin thread some time to process remaining messages before it is stopped or even "killed" when the process exits. What should it return?

Reply:
it's a kind of design question, whether the calls should return anything to at least give the opportunity to handle any result or side effect. if there are no side effects to be concerned, then it doesn't need anything

Reply:
Currently, I try to handle everything within the PluginHandler. Is there anything that I missed and that the external code could or should do?

Reply:
it's OK in either way. but it should be made somewhat explicit, if any action or effect is expected of the call -- and whether something depends on it. e.g. like "fire and forget" vs "accounting"